### ISSUE TITLE
API compatibility of DirichletBC

### ISSUE THREAD

**Issue by APaganini** [firedrakeproject/firedrake#957]
In dolfin it seems you can write

`bc = DirichletBC(V, "-1.0", "on_boundary")`

On firedrake this minimal example
```
from firedrake import *
mesh = UnitSquareMesh(4, 4)
V = FunctionSpace(mesh, "CG", 3)
bc = DirichletBC(V, "-1.0", "on_boundary")
```

returns

```
/Volumes/Scratch/Users/paalbert/Documents/FIREDRAKE/firedrakeAdjoint/src/firedrake/firedrake/interpolation.pyc in make_interpolator(expr, V, subset)
    105     if numpy.prod(expr.ufl_shape, dtype=int) != sum(dims):
    106         raise RuntimeError('Expression of length %d required, got length %d'
--> 107                            % (sum(dims), numpy.prod(expr.ufl_shape, dtype=int)))
    108 
    109     if not isinstance(expr, firedrake.Expression):

RuntimeError: Expression of length 1 required, got length 4
```

**Comment by wence-** [comment]
This happens because in the `DirichletBC` constructor we call `expression.to_expression` on the provided value, which, because strings are iterable, turns it into `Expression(["-", "1", ".", "0"])`.  It should be straightforward to handle this case in `to_expression`, would you like to have a go?

**Comment by APaganini** [comment]
I can give it a try, but have to think how to handle the worst case
`DirichletBC(V, ["0", "1"], "on_boundary")`

**Comment by wence-** [comment]
That already works, I think (you can debug this in isolation by checking what the result of `expressions.to_expression(["0", "1"])` is).

**Comment by APaganini** [comment]
ok, I think that changing
```
def to_expression(val, **kwargs):
    """Convert val to an :class:`Expression`.

    :arg val: an iterable of values suitable for a code snippet in an
         :class:`Expression`.
    :arg \*\*kwargs: keyword arguments passed to the
         :class:`Expression` constructor (which see).
    """
    try:
        expr = ["%s" % v for v in val]
    except TypeError:
        # Not iterable
        expr = "%s" % val

    return Expression(code=expr, **kwargs)
```

into

```
def to_expression(val, **kwargs):
    """Convert val to an :class:`Expression`.

    :arg val: an iterable of values suitable for a code snippet in an
         :class:`Expression`.
    :arg \*\*kwargs: keyword arguments passed to the
         :class:`Expression` constructor (which see).
    """
    if isinstance(val, str):
        expr = [val]
    else:
        try:
            expr = ["%s" % v for v in val]
        except TypeError:
            # Not iterable
            expr = "%s" % val

    return Expression(code=expr, **kwargs)
```

should do the job. Do you suggest a more elegant way?

**Comment by wence-** [comment]
Please go ahead and propose the change.  Then we can review it.

**Comment by wence-** [comment]
Fixed by #961.
